apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/MethodHierarchyMatcher.java

```diff
@@ -39,45 +45,60 @@ public class MethodHierarchyMatcher extends ElementMatcher.Junction.AbstractBase
-        this(extraMethodMatcher, not(is(TypeDescription.ForLoadedType.OBJECT)));
+        this(extraMethodMatcher, not(is(TypeDescription.ForLoadedType.OBJECT)), any());
     }
 
-    private MethodHierarchyMatcher(ElementMatcher<? super MethodDescription> extraMethodMatcher, ElementMatcher<? super TypeDescription> superClassMatcher) {
+    private MethodHierarchyMatcher(ElementMatcher<? super MethodDescription> extraMethodMatcher, ElementMatcher<? super TypeDescription> superClassMatcher, ElementMatcher<? super TypeDescription> hierachyMatcher) {
         this.extraMethodMatcher = extraMethodMatcher;
         this.superClassMatcher = superClassMatcher;
+        this.hierarchyMatcher = hierachyMatcher;
+    }
+
+    public MethodHierarchyMatcher onSuperClassesThat(ElementMatcher<? super TypeDescription> superClassMatcher) {
+        return new MethodHierarchyMatcher(extraMethodMatcher, superClassMatcher, hierarchyMatcher);
     }
 
-    public ElementMatcher<MethodDescription> onSuperClassesThat(ElementMatcher<? super TypeDescription> superClassMatcher) {
-        return new MethodHierarchyMatcher(extraMethodMatcher, superClassMatcher);
+    public MethodHierarchyMatcher whereHierarchyContains(ElementMatcher<? super TypeDescription> hierarchyMatcher) {
+        return new MethodHierarchyMatcher(extraMethodMatcher, superClassMatcher, hierarchyMatcher);
     }
 
     @Override
     public boolean matches(MethodDescription targetMethod) {
-        return declaresInHierarchy(targetMethod, targetMethod.getDeclaringType().asErasure());
+        return declaresInHierarchy(targetMethod, targetMethod.getDeclaringType().asErasure(), new ArrayDeque<TypeDescription>());
     }
 
-    private boolean declaresInHierarchy(MethodDescription targetMethod, TypeDescription type) {
-        if (declaresMethod(named(targetMethod.getName())
-            .and(returns(targetMethod.getReturnType().asErasure()))
-            .and(takesArguments(targetMethod.getParameters().asTypeList().asErasures()))
-            .and(extraMethodMatcher))
-            .matches(type)) {
-            return true;
-        }
-        for (TypeDescription interfaze : type.getInterfaces().asErasures()) {
-            if (superClassMatcher.matches(interfaze)) {
-                if (declaresInHierarchy(targetMethod, interfaze)) {
-                    return true;
+    private boolean declaresInHierarchy(MethodDescription targetMethod, TypeDescription type, Deque<TypeDescription> hierarchy) {
+        hierarchy.push(type);
+        try {
+            if (declaresMethod(named(targetMethod.getName())
+                .and(returns(targetMethod.getReturnType().asErasure()))
+                .and(takesArguments(targetMethod.getParameters().asTypeList().asErasures()))
+                .and(extraMethodMatcher))
+                .matches(type)
+             && !new TypeList.Explicit(new ArrayList<>(hierarchy))
+                .filter(hierarchyMatcher)
+                .isEmpty()
```

Comment:
[minor] this one probably deserves to be explained a bit, maybe through a comment. If I understand it correctly it's "the type matches the hierarchyMatcher", maybe nesting an IF in this case could help.